Mathlib/Analysis/SpecialFunctions/MulExpNegMulSq.lean

```diff
@@ -0,0 +1,216 @@
+  simp only [differentiableAt_id', differentiableAt_expNegSq, DifferentiableAt.mul]
+
+theorem differentiable_mulExpNegSq : Differentiable ℝ (fun x => x * exp (-x * x)) :=
+  fun _ => differentiableAt_mulExpNegSq _
+
+theorem hasDerivAt_mulExpNegSq (y : ℝ) :
+    HasDerivAt (fun x => x * exp (-x * x)) (exp (-y * y) + y * (exp (-y * y) * (-2 * y))) y := by
+  nth_rw 1 [← one_mul (exp (-y * y))]
+  exact HasDerivAt.mul (hasDerivAt_id' y) (hasDerivAt_expNegSq y)
+
+theorem deriv_mulExpNegSq (y : ℝ) : deriv (fun x => x * exp (- x * x)) y =
+    exp (-y * y) + y * (exp (-y * y) * (-2 * y)) :=
+  HasDerivAt.deriv (hasDerivAt_mulExpNegSq y)
+
+theorem deriv_mulExpNegMulSq_one (x : ℝ) : deriv (mulExpNegMulSq 1) x =
+    exp (-x * x) + x * (exp (-x * x) * (-2 * x)) := by
+  rw [← deriv_mulExpNegSq]
+  exact EventuallyEq.deriv_eq (Eventually.of_forall (mulExpNegSq_one))
+
+theorem nnnorm_deriv_mulExpNegMulSq_one_le_one (x : ℝ) : ‖deriv (mulExpNegMulSq 1) x‖₊ ≤ 1 := by
+  rw [← NNReal.coe_le_coe, coe_nnnorm, norm_eq_abs, deriv_mulExpNegMulSq_one]
+  have heq : exp (-x * x) + x * (exp (-x * x) * (-2 * x)) = exp (-x * x) * (1 + 2 * -x * x) := by
+    ring
+  rw [heq, abs_mul, abs_exp]
+  set y := x * x with hy
+  rw [neg_mul, mul_assoc, neg_mul, ← hy]
+  apply mul_le_of_le_inv_mul₀ (zero_le_one' ℝ) (exp_nonneg _)
+  simp [← exp_neg (-y), abs_le]
+  constructor
+  · have twomul_le_exp : 2 * y ≤ exp y := by
+      apply le_trans _ exp_one_mul_le_exp
+      have : 2 ≤ exp 1 := by
+        apply le_of_lt (lt_trans _ exp_one_gt_d9)
+        norm_num
+      apply mul_le_mul_of_nonneg_right this (mul_self_nonneg x)
+    exact le_trans twomul_le_exp ((le_add_iff_nonneg_right (exp y)).mpr zero_le_one)
+  · apply le_trans (one_le_exp (mul_self_nonneg x)) (le_add_of_nonneg_right _)
+    simp [hy, (mul_self_nonneg x)]
+
+theorem lipschitzWith_one_mulExpNegMulSq_one : LipschitzWith 1 (mulExpNegMulSq 1) :=
+  lipschitzWith_of_nnnorm_deriv_le (Differentiable.mul differentiable_id' (by simp))
+    nnnorm_deriv_mulExpNegMulSq_one_le_one
+
+end mulExpNegMulSq
+
+section mulExpNegMulSq_comp
+
+/-! ### Properties of `(mulExpNegMulSq ε).comp g` -/
+
+variable {E : Type*} {g : E → ℝ} {ε : ℝ} {x : E}
+
+/--
+`mulExpNegMulSq ε` transforms a function `g` into another function with useful
+boundedness and convergence properties.
+-/
+
+theorem mulExpNegMulSq_eq_sqrt_mul_mulExpNegMulSq_one (hε : ε > 0) :
+    (mulExpNegMulSq ε).comp g x
+    = ε.sqrt⁻¹ * mulExpNegMulSq 1 (sqrt ε * g x) := by
+  simp [mulExpNegMulSq]
+  have h : ε.sqrt * g x * (ε.sqrt * g x) = ε * g x * g x := by
+    ring_nf
+    rw [sq_sqrt hε.le, mul_comm]
+  rw [h, eq_inv_mul_iff_mul_eq₀ _]
+  · ring_nf
+  · intro h
+    rw [← pow_eq_zero_iff (Ne.symm (Nat.zero_ne_add_one 1)), sq_sqrt hε.le] at h
+    linarith
+
+/-- if `ε > 0`, then `(mulExpNegMulSq ε).comp g` is bounded by `ε.sqrt⁻¹` -/
+theorem abs_mulExpNegMulSq_comp_le (hε : ε > 0) : |(mulExpNegMulSq ε).comp g x| ≤ ε.sqrt⁻¹ := by
```

Comment:
The standard writing, here and elsewhere in this file, should be `(mulExpNegMulSq ε) ∘ g` instead of `(mulExpNegMulSq ε).comp g`.